### PR TITLE
Corrected logic in the SchemaBase.Status

### LINKED ISSUE #73: Object not set... error when generating differences
When generating differences between databases, I get the following issue:

```
1.  To report an error search first in the Github issues to see if it's already been reported.
2a. If there is no issue, click 'New issue' and paste the error details
    into the body of the issue. To copy the error press "Copy error"
    (At least email the details to opendbdiff@gmail.com)

2b. If the issue exists, but your situation is different please leave a comment with the details.

•   If possible, please attach a SQL script creating the two databases with
    the minimum necessary to reproduce the problem.

Error information
Error Comparing Databases (while Reading Text Objects: [DocumentInsert])
NullReferenceException: Object reference not set to an instance of an object.
   at OpenDBDiff.Schema.Model.SchemaBase.set_Status(ObjectStatus value) in OpenDBDiff.Schema\Model\SchemaBase.cs:line 239
   at OpenDBDiff.Schema.Model.SchemaBase.set_Status(ObjectStatus value) in OpenDBDiff.Schema\Model\SchemaBase.cs:line 239
   at OpenDBDiff.Schema.Model.SchemaBase.set_Status(ObjectStatus value) in OpenDBDiff.Schema\Model\SchemaBase.cs:line 239
   at OpenDBDiff.Schema.SQLServer.Generates.Compare.CompareColumns.GenerateDifferences[T](Columns`1 originFields, Columns`1 destinationFields) in OpenDBDiff.Schema.SQLServer.Generates\Compare\CompareColumns.cs:line 81
   at OpenDBDiff.Schema.SQLServer.Generates.Compare.CompareTables.DoUpdate[Root](SchemaList`2 originFields, Table node) in OpenDBDiff.Schema.SQLServer.Generates\Compare\CompareTables.cs:line 14
   at OpenDBDiff.Schema.SQLServer.Generates.Compare.CompareBase`1.GenerateDifferences[Root](SchemaList`2 originFields, SchemaList`2 destinationFields) in OpenDBDiff.Schema.SQLServer.Generates\Compare\CompareBase.cs:line 54
   at OpenDBDiff.Schema.SQLServer.Generates.Compare.CompareDatabase.GenerateDifferences(Database origin, Database destination) in OpenDBDiff.Schema.SQLServer.Generates\Compare\CompareDatabase.cs:line 21

86EA0D2CEBDC074C8D1D3FED34E60F8D
```

It's strange because it is the `Parent` property that is null even though it is inside a `Parent != null` if statement... so that leads me to think it's a threading issue but haven't been able to prove it as yet.

### PR DESCRIPTION
**Pull request by MarkGaze**
The current logic is checking that the Parent is not null but then the ORs allow the code to continue, I'm assuming that the last three statements were meant to be grouped. Can you check the validity of this @igitur? I ran it and it fixed #73 for me.

### CODE REVIEW

**Comment by igitur** [comment]
I want to reproduce the problem on my side before accepting any PRs.

**Comment by MarkGaze** [comment]
I've added a unit test to prove the fix, if you change back to the old code you can recreate the issue I had on my machine which is fixed by the code change

**Comment by igitur** [comment]
I'm not convinced by this PR. One last time... please attach schema creation scripts for objects that reproduces this problem.

**Comment by MarkGaze** [comment]
I don't know which of the thousands of objects that are in my databases are creating the issue, and I haven't got the time to spend searching for them otherwise I would gladly send them.

It seems fairly clear that the logic in that **one line of code** I changed in the original commit is incorrect, as it allows the execution of the code inside it regardless of whether the Parent is null or not and that code contains usages of Parent.Status on all other branches of code.

The unit test that I added proves this further.

So, either the fix I made (which fixed both my problem and the unit test I wrote that recreated the issue) is correct or the if statement needs rethinking.

I'm evaluating this tool for use in my workplace, which has a lot of large databases so I expected problems but the open nature of the codebase is a major benefit to me and I'm more than happy to contribute to it but having to prove each fix with scripts is going to be a problem which is why I added the unit test that performs the same function, if there are other critical paths that would use this code then I'm happy to discuss and add unit tests for them too.

**Comment by igitur** [comment]
You don't have to go through thousands of objects. The error message in #73 states that the culprit object is `DocumentInsert`.

`Error Comparing Databases (while Reading Text Objects: [DocumentInsert])`.

Unfortunately the current unit test coverage is extermely poor, so adding a unit test to substantiate your PR is not enough. It is very possible that this PR could break other areas while passing the unit test that you added. Therefore I'm not going to blindly accept code changes without understanding the actual cause of the problem.

**Comment by MarkGaze** [comment]
Except (and this is my fault, because I removed the comment with the files in which had this information) the file changed to a different one after I logged the issue. I'll have another look.

File: OpenDBDiff.Schema/Model/SchemaBase.cs [hunk 1]
```diff
@@ -229,7 +229,7 @@ public virtual ObjectStatus Status
                     status = value;
 
                 // Si el estado de la tabla era el original, lo cambia, sino deja el actual estado.
-                if (Parent != null && Parent.Status == ObjectStatus.Original || value == ObjectStatus.Rebuild || value == ObjectStatus.RebuildDependencies)
+                if (Parent != null && (Parent.Status == ObjectStatus.Original || value == ObjectStatus.Rebuild || value == ObjectStatus.RebuildDependencies))
```
Comment (igitur) [inline OpenDBDiff.Schema/Model/SchemaBase.cs]: I agree with this fix, but I'd like to rework the logic in this property a bit. But I still want you to get the credit. Would you mind changing the entire `SchemaBase.Status` property to:

```c#
        public virtual ObjectStatus Status
        {
            get { return status; }
            set
            {
                if (status != ObjectStatus.Rebuild && status != ObjectStatus.RebuildDependencies)
                    status = value;

                if (Parent == null) return;

                // Si el estado de la tabla era el original, lo cambia, sino deja el actual estado.
                // If the state of the table was the original, it changes it, but leaves the current state. (Google translated)
                if (Parent.Status == ObjectStatus.Original
                    || value == ObjectStatus.Rebuild
                    || value == ObjectStatus.RebuildDependencies)
                {
                    switch (value)
                    {
                        case ObjectStatus.RebuildDependencies:
                        case ObjectStatus.Rebuild:
                            Parent.Status = value;
                            break;

                        case ObjectStatus.Original:
                            break;

                        default:
                            Parent.Status = ObjectStatus.Alter;
                            break;
                    }
                }
            }
        }
```

File: OpenDBDiff.sln [hunk 2]
```diff
@@ -27,6 +27,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenDBDiff.Schema.SQLServer
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenDBDiff.Schema.SQLServer.GeneratesTests", "OpenDBDiff.Schema.SQLServer.GeneratesTests\OpenDBDiff.Schema.SQLServer.GeneratesTests.csproj", "{94F5C91A-6268-41E9-A15E-FA5451BB2B6F}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenDBDiff.SchemaTests", "OpenDBDiff.SchemaTests\OpenDBDiff.SchemaTests.csproj", "{9CBBBA4C-AA7E-41DE-B277-2F5CE7D31ACC}"
```
Comment (igitur) [inline OpenDBDiff.sln]: Revert, as discussed.

**Comment by MarkGaze** [comment]
@igitur I think we should be good now!

**Comment by jspraul** [comment]
Much thanks to @MarkGaze for doing the exra work to polish this up per @igitur's guidance.

**Comment by igitur** [comment]
Thanks for the contribution.

**Pull request merged**